Lazily deserialize cache entries

jonathanhefner opened the pull request:
This adds a cache optimization such that expired and version-mismatched cache entries can be detected without deserializing their values.  This optimization is enabled when using cache format version >= 7.1 or a custom serializer.

jonathanhefner commented on activesupport/lib/active_support/cache/serializer_adapter.rb:
```diff
@@ -0,0 +1,171 @@
+
+module ActiveSupport
+  module Cache
+    class SerializerAdapter # :nodoc:
```
`SerializerAdapter` is a coder with swappable serializer and compressor.

Is the primary concern integration with something like paquito?

What if we (1) rename `SerializerAdapter` to `Cache::Coder`, and (2) support specifying either a `:coder` option _OR_ `:serializer` and `:compressor` options for `Cache::Store`?  When `:coder` is specified, it will be used directly, no `Cache::Coder` involved.  Otherwise, `:serializer` and `:compressor` will be used via `Cache::Coder`, with their default values controlled by the cache format version.

casperisfine replied:
> Is the primary concern integration with something like paquito?

In part yes. I think if a `coder` is specified, it should have total control over the serialization, Active Support shouldn't prepend anything to the provided payload etc.

>  support specifying either a :coder option OR :serializer and :compressor options

I think you touch on something here.

Paquito went the route of taking total control over how the `Entry` (and its content) is serialized.

You can't combine that with a configurable compressor. But in many case users may only want to control how the `Entry#value` is serialized.

If both were explicitly distinct I think it would make the code clearer.

So 👍 to experiment that (and sorry for dragging my feet on this, I know you've put a lot of efforts into it, I'm just super worried about the maintainability of all this).

jonathanhefner committed:
Lazily deserialize cache entries

This adds a cache optimization such that expired and version-mismatched
cache entries can be detected without deserializing their values.  This
optimization is enabled when using cache format version >= 7.1 or a
custom serializer.

Co-authored-by: Jean Boussier <jean.boussier@gmail.com>

byroot commented on activesupport/lib/active_support/cache/coder.rb:
```diff
@@ -88,6 +90,28 @@ def load(dumped)
+
+          def value
+            if @type
+              @value = @coder._load_payload(@value, @type)
```
Not a big fan of that `_load_payload`.

Could we pass both the `serializer` and `compressor`? If the payload isn't compressed, we pass `compressor = nil`.

How does that sound?

jonathanhefner replied:
If that is what you prefer, we can do that.

The reason I organized it as `_load_payload` is the same reason I originally used a thunk: this code is essentially part of `load`, it is merely delayed.  Encapsulating the code in `_load_payload` allows it to be near to `load` so they can be understood together.

I've pushed a commit that passes `serializer` and `compressor`, and `compressor = nil` when the payload isn't compressed.  In my opinion, the code is disjointed and less clear.

If you can tell me what you dislike about `_load_payload`, perhaps I can iterate on it.  If you dislike the name, I can change it.  If you dislike `LazyEntry` delegating to `@coder._load_payload`, I can pass (a cached copy of) `method(:_load_payload)` instead.

byroot replied:
> If you can tell me what you dislike about `_load_payload`,

A few things:

  - It increase the interface, particularly with something that is public but private `_load`.
  - I like that coder, serializers etc have a simple and standard `load/dump`, `inflate/deflate` API.
  - It makes the entry and coder reflation two way, I prefer when reference can go only in one direction.

Looking at your last commit now.

jonathanhefner replied:
> * It increase the interface, particularly with something that is public but private `_load`.
>
> * I like that coder, serializers etc have a simple and standard `load/dump`, `inflate/deflate` API.

These could be solved by passing something like `method(:_load_payload)` instead of the coder instance.

> * It makes the entry and coder reflation two way, I prefer when reference can go only in one direction.

In terms of `LazyEntry` knowing about `Cache::Coder`, passing `method(:_load_payload)` would also abstract that relationship.

In terms of references and lifetimes, `LazyEntry` must still hold references to call `@serializer.load` and `@compressor.inflate`, so there isn't much difference either way.

byroot commented on activesupport/lib/active_support/cache/coder.rb:
```diff
@@ -88,6 +81,36 @@ def load(dumped)
+            if @type
+              @value = @compressor.inflate(@value) if @compressor
+
+              if string_encoding = STRING_ENCODINGS[@type & ~COMPRESSED_FLAG]
+                @value.force_encoding(string_encoding)
+              else
```
Ok, so I think if we extracted this to be a serializer, it would be perfect.

So basically if we had something like:

```ruby
module Utf8Serializer
  extend self
  def load(payload)
    payload.force_encoding(Encoding::UTF_8)
  end
end
STRING_SERIALIZERS = {
  0x02 => Utf8Serializer,
}  
```

This way you can check the type before calling `LazyEntry.new` and always pass a serializer, and don't need to pass the type.

jonathanhefner replied:
I've pushed two more commits (031a4e9d0d9070728d3ee80804ab4553afd41598 and b819095cf5fa1a30c05757344f79c0379c03dfe4).  The first implements `Utf8Serializer`, etc.  That seemed too cumbersome to me, so the 2nd commit refactors them as `StringDeserializer` instances.

It's worth pointing out that their `load` method mutates its argument, whereas typical `load` methods do not.  That does not affect our implementation, but it is a leaky abstraction.

byroot replied:
> so the 2nd commit refactors them as StringDeserializer instances.

Yes, good call, it's quite nice.

> It's worth pointing out that their load method mutates its argument,

That's the case too with the existing `Coder`. it's indeed a leaky abstraction, but here we know we the caller won't use the string, so it's a nice little perf gain.

byroot commented on activesupport/lib/active_support/cache/coder.rb:
```diff
@@ -88,6 +81,36 @@ def load(dumped)
+
+          def value
+            if @type
+              @value = @compressor.inflate(@value) if @compressor
```
I have a small worry here. If the serializer later raise for some reason, the lazy entry will be less with a decompressed `@value`.

I think we should store the decompressed payload in a local variable and only re-assing `@value` on both decompression and deserialization succeeded.

byroot commented on activesupport/lib/active_support/cache/coder.rb:
```diff
@@ -56,15 +56,10 @@ def load(dumped)
 
-        payload = @compressor.inflate(payload) if type & COMPRESSED_FLAG > 0
+        compressor = @compressor if type & COMPRESSED_FLAG > 0
+        serializer = STRING_DESERIALIZERS.fetch(type & ~COMPRESSED_FLAG) { @serializer }
```
```suggestion
        serializer = STRING_DESERIALIZERS.fetch(type & ~COMPRESSED_FLAG, @serializer)
```

nitpick / micro-opt might as well save the block call.
